Extend and use new CRYPTO_THREAD_[get|set]_local api to reduce our reliance on OS thread-local variables

nhorman opened the pull request:
Recently, as part of https://github.com/openssl/openssl/issues/27757, we found that:
a) We are using a lot of thread-local storage keys (about 6, or 12 if you load the fips provider)
b) We create thread-local storage in a few places per-context

(a) is somewhat manageable, but it would be nice if we didn't consume so many, as its a limited resource at the OS interface, and we should play nice with applications using openssl in this regard

(b) is a real problem, as it creates an implicit limit on how many libctx objects we can create.  By my testing, we currently create up to 4 new thread-local keys per context (in the fips provider we create an rcu key, two keys in the rand library, and a thread event handler key).  Linux for instance only allows 512 keys per process, and so as a result, if we create all of those keys on every libctx allocation, we are implicitly limited to only creating 512/4 - (6 keys created globally) = 126 libctx objects.  That won't always be the case depending on our configuration, but it would be easy for us to exhaust that limit.

So lets try to improve the situation.  This PR creates the internal api calls CRYPTO_THREAD_[get|set]_local_ex.  These apis create a single master key at the OS level for thread local storage, which points to a series of sparse arrays which then hold the actual per thread*ctx data pointers that we want to look up.  Each thread indexes the table via two values:
1) The key (a fixed ennumeration of keys we support via this api which can be extended as needed)
2) The library context

The remainder of this PR adapts users of the legacy CRYPTO_THREAD_[get|set]_local api to use this new api, reducing our overall OS level key usage, and eliminating the per-context allocation of an OS level key in for those specific offenders.

fochoao-dev commented:
Boolean? The main key is hidden outside of public_html, then the public key

t8m commented:
This looks like a valid approach, except for one thing - what about the performance of using the sparse array as the backend storage. Would it make sense to keep using directly thread locals at least for the default context thread local? That is there just once per libcrypto initialization.

nhorman commented:
I've been trying to ignore any possible performance issues while this is in draft, just to make sure that the new api functionally works.  Once it all works, I think it makes sense to run some tests and have a debate around which keys it makes sense to manage internally vs which we should use the backing system thread-local infrastructure.

nhorman commented:
Fixed up my stats collection to get better results.

Using two sparse arrays produces this comparison:
![thread_local_speed_sa](https://github.com/user-attachments/assets/993f206b-84de-4077-a122-d8f516c44478)

And replacing the top level sparse array with a fixed size array, allocated based on the number of key in our key id ennumeration:
![thread_local_speed_fa](https://github.com/user-attachments/assets/d80d1e09-b299-46fc-82ad-612f04652456)


does show a little improvement, but not a significant amount.  Given these results I would say we should just leave both sparse arrays in place and make the judgement call as to what keys to move into this api based on that.  So far as I can tell the only keys that have to move (lest we limit ourselves to the number of contexts we want to be able to create) are:
1) The rcu key
2) The two drbg keys
3) The thread local event handler key

Everything else is negotiable based on whats easier to use.

t8m commented:
There is one more possible radical way how to avoid per-context thread locals (and all the thread locking stuff as well!). Basically have a boolean in an OSSL_LIB_CTX indicating the library context is thread-specific. I.e. it will never be used with multiple threads simultaneously. You can then drop all the per-libctx thread locals and locks completely.

t8m commented:
The reasoning for the above - with it you can have basically two application models:
1. shared libctxes, new libctx is used only for either some special operation like fips-enabled operation or to separate library usage of openssl among unrelated 3rd party libraries and the application.
2. per-thread libctxes - to separate libcrypto/libssl data per-thread

Both models have its own overheads and are suitable for different applications.

nhorman commented:
@t8m hmm, I think I could see how that might work, but I'd be concerned about enforcing the behavior (If I understand it correctly).  To elaborate, if we create a library context with OSSL_LIBCTX_new that is "thread specifc", we could avoid creating a thread-local storage space for it, opting instead to just have libctx based storage for any such data.  But in that event, how do we guarantee that the thread referencing that data is the thread that created it?  In other words, if we encounter an error in which thread A created a libctx, but then thread B somehow references it, how do we detect that error, and what do we do in response to it?

mattcaswell commented:
> Both models have its own overheads and are suitable for different applications.

It's an interesting approach. But i see it as orthogonal to this PR.

> if we encounter an error in which thread A created a libctx, but then thread B somehow references it, how do we detect that error, and what do we do in response to it?

We do not detect or respond to it. That would be a user error IMO.

Sashan commented:
> > Both models have its own overheads and are suitable for different applications.
> 
> It's an interesting approach. But i see it as orthogonal to this PR.
> 
> > if we encounter an error in which thread A created a libctx, but then thread B somehow references it, how do we detect that error, and what do we do in response to it?
> 
> We do not detect or respond to it. That would be a user error IMO.

I think the approach of thread-friendly libctx and single-thread libctx is like adding yet another knob user/application can use to hurt its knee. If user/vendor is absolutely sure not to use threads in libcrypto then the first option is to build library without support for threads. It's just my opinion.

nhorman commented:
@t8m @paulidale @Sashan 
I think its something of a moot point.  I thought @t8m 's idea was interesting, so I put together this branch last night:
https://github.com/nhorman/openssl/tree/singlethreaded-contexts

Which creates a context aware read/write lock api, and a ctx ownership api.  It allows for the actual aquisition/release of rwlocks to be skipped in the event that a context is marked as being used by a single thread.  The branch above also updates all of our single threaded tests to be marked as being such so we avoid unneeded locking.  Then I ran our test suite on both the above branch and the head of master to compare run times as compiled with:
```
Configure enable-fips
```
The result was that there was no speedup observed.

Additionally, I modified our perftools handshake test, which allows for the creation of a libctx per thread, and modified it to optionally allow us to set exclusive thread ownership.  And got these results:
| threads | handshakes/sec (shared)  | handshake/sec (exclusive) | % improvement |
|-------------|-----------------------------------|-------------------------------------|----------------------|
|1| 12271.939165 | 11593.465414| |-5.80%
|10 |11603.953300 |11637.672848| 0.28%|
|20 |11853.849832 |11620.306101| -2.00%|
|30 |11451.414170 |11489.920142| .33% |
|40 |11900.656669 |11530.158311| -3.21%|
|50 |11445.225629 |11459.702557| 0.12%|
|60 |11727.501180 |11522.610652| -1.7%|
|70 |11373.721021 |11447.326729| 0.64%|
|80 |11738.381891 |11456.176459| -2.4%|
|90 |11417.584462 |11519.459030| 0.88%|
|100 | 11475.396576 |11326.195453| -1.31%|

So again, we see no marked improvement.  If anything things get just a tiny bit worse.

What I would conclude from the above is that, while avoiding locking helps in an environment where data is shared, avoiding locking in an uncontended case (i.e. where we are guaranteed to have exclusive access to the data we protect), has little to no impact, as the cpu cycles required to take an uncontested lock are minimal.  In fact the effort to check an extra conditional to determine if the lock needs to be taken outweighs the work needed to just take the lock in the first place.

And while we could expand on this to move from thread-local storage to context specific storage, I'd be concerned that the extra checking we would need to do to determine if we need to use thread-local storage or not would outpace the cost of just unilaterally doing the thread-local lookup.

Its an unfortunate result, as I thought it was a neat idea, but cest-la-vie.

I should also not here, this is all orthogonal to the Pr attached to this issue, which addresses the very real problem of thread-local storage keys exhaustion.

paulidale commented:
Another possible source of parallel slowness would be our atomic operations.  On recent processors they are similar in cost to a cache miss and we have started using them a lot more.

npajkovsky commented:
Tested against the 52dba1c098d9b0388afa4db2fbe3461df67461c6 openssl branch.

`-s` disables SSL_CTX sharing, which means that each handshake creates its own SSL_CTX.
`-p` uses per thread libctx
`-P` uses a libctx pool of size 16, distributing libctx instances to threads using a simple round-robin approach.

the `handsake` binary has been tested on M4Pro. The values in the cells are handshakes/sec.
```
| Threads | NoOpts |    -p |   -s | -s -P |    -P | -s -p |
|---------+--------+-------+------+-------+-------+-------|
|       1 |   1736 |  1658 | 1042 |  1022 |  1660 |  1072 |
|      10 |  11994 | 12767 | 6766 |  7408 | 12919 |  7964 |
|      20 |  12304 | 13179 | 5652 |  5621 | 13339 |  5689 |
|      30 |  12123 | 13374 | 5499 |  5151 | 13204 |  5049 |
|      40 |  11667 | 13587 | 5370 |  4898 | 13231 |  5042 |
|      50 |  12209 | 13445 | 5406 |  4855 | 13273 |  5017 |
|      60 |  12311 | 13409 | 5213 |  4857 | 13389 |  4998 |
|      70 |  12072 | 13244 | 5092 |  4721 | 13170 |  4748 |
|      80 |  11961 | 13205 | 5112 |  4640 | 13010 |  4804 |
|      90 |  11897 | 13098 | 5075 |  4477 | 12670 |  4721 |
|     100 |  12074 | 12384 | 4937 |  4458 | 13229 |  4677 |
|     110 |  12200 | 12815 | 4858 |  4571 | 13097 |  4530 |
|     120 |  11838 | 12735 | 4763 |  4501 | 12589 |  4529 |
|     130 |  11079 | 12505 | 4652 |  4548 | 12611 |  4348 |
|     140 |  11314 | 12701 | 4517 |  4324 | 12620 |  4393 |
|     150 |  11245 | 12396 | 4419 |  4425 | 12554 |  4438 |
|     160 |  11048 | 12427 | 4367 |  4296 | 12904 |  4300 |
|     170 |  11139 |       | 4330 |  4357 | 12579 |       |
|     180 |  10399 |       | 4140 |  4228 | 12654 |       |
|     190 |  11275 |       | 4004 |  4183 | 12595 |       |
|     200 |  11297 |       | 4093 |  3644 | 12379 |       |
```

Sashan commented on crypto/initthread.c:
```diff
@@ -121,6 +123,32 @@ init_get_thread_local(CRYPTO_THREAD_LOCAL *local, int alloc, int keep)
 
+#else
+static THREAD_EVENT_HANDLER **
+init_get_thread_local_ex(OSSL_LIB_CTX *ctx, int alloc, int keep)
```
We currently call `init_get_thread_local()` with two forms:
  - `init_get_thread_local(ctx, 1, 0);`
  - `init_get_thread_local(ctx, 0, 0);`

it feels like arg2 (`keep`) is always zero. can the argument be removed completely?

nhorman replied:
ossl_ctx_thread_start  calls  init_get_thread_local(ctx, 1, 0) in the non-fips case.

Sashan replied:
I see where the confusion is coming from here. The `keep` argument at `init_get_thread_local_ex()` is always 0, therefore, I think it can be removed for `_ex()` variant.

Sashan replied:
To be honest I don't like this _swiss army knife_ kind of function. I would prefer to have more functions with proper name. For example:
   init_get_thread_local(&x, 0, 0); => get_thread_local(&x);
   init_get_thread_local(&x, 1, .); => create_thread_local(&x);
   init_get_thread_local(&x, ., 1); => destroy_thread_local(&x);
the 1,0 arguments make reading the code lot harder. This is just suggestion I don't insist on.

t-j-h replied:
> To be honest I don't like this _swiss army knife_ kind of function. 

Or if you want a single function they have flags with macros for their values (or enums if internal) so it is clearer.

init_get_thread_local(&x, OSSL_THREAD_LOCAL_GET); 
init_get_thread_local(&x, OSSL_THREAD_LOCAL_CREATE); or _INIT?
init_get_thread_local(&x, OSSL_THREAD_LOCAL_DESTROY);

It also allows for more "operations" to be added without adding more functions or needing to add "ex" functions etc.

t8m replied:
I'm with @Sashan on this one - I'd just create 3 functions and use them as appropriate. These are internal functions so there is no problem with creating new one or changing semantics of existing one.

Sashan commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+    /*
+     * Lastly assign our data pointer
+     */
+    ctxd->ctx_data = data;
```
Shall we fail if `ctxd->ctx_data` is not `NULL` so we can give a caller chance to avoid a memory leak (accident overwrite of existing key)? Perhaps more changes would be required if we will go this way.

nhorman replied:
Normally the call pattern to update a thread-local pointer looks like this:
```
tmp = CRYPTO_THREAD_get_local[_ex](key, [ctx]);
CRYPTO_THREAD_set_local[_ex](key, [ctx], newptr);
OPENSSL_free(tmp);
```

If we were to introduce a failure on non-null then the calling pattern would have to be something like:
```
tmp = CRYPTO_THREAD_get_local[_ex](key, [ctx]);
CRYPTO_THREAD_set_local[_ex](key, [ctx], NULL);
CRYPTO_THREAD_set_local[_ex](key, [ctx], newptr);
OPENSSL_free(tmp);
```
which seems less efficient.

Given that asan will catch any leaks, I think we're ok as we are on this account.

Sashan commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+ * Data Design:
+ *
+ * per-thread master key data->  ┌──────────────┬─┐
+ *                               │              │ │
```
this is not what I'm seeing when opening the code in non-utf8 terminal please check file encoding when you save the file.

Sashan commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+ *   - CRYPTO_THREAD_get_local_ex() retrieves data associated with a key and
+ *     context.
+ *   - CRYPTO_THREAD_set_local_ex() associates data with a given key and
+ *     context, allocating tables as needed.
```
I think we will need to add explicit _create() and _destroy() functions. This will allow function _set() to fail if we attempt to overwrite value at existing key.

Sashan commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+DEFINE_SPARSE_ARRAY_OF(MASTER_KEY_ENTRY);
+
+/**
+ * @brief holds our per therad data with the operating system
```
s/therad/thread

Sashan commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+/**
+ * @brief Informs the library if the master key has been set up
+ *
+ * State variable to track if we have initalized the master_key
```
s/initalized/initialized

Sashan commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+ *
+ * State variable to track if we have initalized the master_key
+ * If this isn't set to 1, then we need to skip any cleanup
+ * in CRYPTO_THREAD_clean_for_fips, as the uninitalized key
```
s/uninitalized/uninitialized

Sashan commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+/**
+ * @brief gate variable to do one time init of the master key
+ *
+ * Run once gate for doing one-time initalization
```
s/initalization/initialization

Sashan commented on crypto/threads_pthread.c:
```diff
@@ -292,29 +294,27 @@ static struct rcu_qp *get_hold_current_qp(struct rcu_lock_st *lock)
 
     OPENSSL_free(data);
-    CRYPTO_THREAD_set_local(lkey, NULL);
+    CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_RCU_KEY, ctx, NULL);
```
Can you swap lines 300 and 299? so `OPENSSL_free()` follows `_set_local_ex(NULL)`?
thanks.

Sashan commented on crypto/threads_win.c:
```diff
@@ -221,29 +222,28 @@ static ossl_inline struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+
     OPENSSL_free(data);
-    CRYPTO_THREAD_set_local(lkey, NULL);
+    CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_RCU_KEY, ctx, NULL);
```
can we swap this line with `OPENSSL_free()` so we set `NULL` first and then call free? thanks.

Sashan reviewed (changes requested):
I think this is heading in the right direction.

Sashan commented:
![vim](https://github.com/user-attachments/assets/b0deca23-95d2-4eba-bc6c-59d3ae881cdf)
this is the screenshout which shows utf-8 chars in non-utf8 terminal.

mattcaswell commented:
> Basically have a boolean in an OSSL_LIB_CTX indicating the library context is thread-specific. I.e. it will never be used with multiple threads simultaneously

An alternative suggestion is to just create a completely lockless `OSSL_LIB_CTX`. All loading/unloading of providers must only be done before the libctx can be used by multiple threads. Probably we would also have to prefetch all the algorithms too. Once all the providers are loaded the libctx could be used like any other (including shared between threads), but no locks would be required.

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+ * and an `OSSL_LIB_CTX *` context.
+ *
+ * NOTE: This differs from the CRYPTO_THREAD_[get|set]_local api set in that
+ * this api stores a single OS level thread-local key per-thread, and manages
```
It's not a single thread-local key per thread (the whole point is that there is one key, but each thread gets different local data) - its per process right?

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+ *                               ┌──────────────┐ │
+ *                               │              │ │
+ *                               │              │ │
+ *                               └──────────────┘ │  sparse  array indexed
```
Not really sure why its a sparse array. We have a fixed number of well known consecutive id values. Surely this should just be a regular array. This is not "sparse" in any way.

nhorman replied:
because it was easier to follow the same pattern for both arrays, but yes, I can make it a fixed size array

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+/*
+ * Define our sparse array functions for MASTER_KEY_ENTRY structs
+ */
+DEFINE_SPARSE_ARRAY_OF(MASTER_KEY_ENTRY);
```
I think this should just be a regular array

mattcaswell replied:
The FIPS provider manages its own thread local data completely separately to libcrypto. The size of the array can be different between the two because they are entirely independent. So if we have a newer libcrypto which no longer users a specific key id that the fips provider uses - it makes no difference. The fips provider is entirely separate anyway.

paulidale replied:
Sparse arrays are not going to save space when only a few indices are missing.  They are implemented as a tree of blocks and any gap under the size of a block will use a full block.

Access is O(log n) where n is the number of elements in the sparse array.  The block size is fairly large so this isn't consequential most of the time.

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+     * stay valid until the thread actually exits, which is what the
+     * clean_master_key function does.  Data held in the sparse arrays
+     * (that is assigned via CRYPTO_THREAD_set_local_ex), are still expected
+     * to be cleaned via the ossl_init_thread_start/stop api.
```
Some optional enhancement in the future, might be to register data specific cleanup handlers. Sometimes `ossl_init_thread_stop` doesn't work and applications can experience leaks (e.g. if OpenSSL is unloaded while threads are still running). At the moment we rely on the user application to call `OPENSSL_thread_stop` for all threads in such a case - which is clunky, not well known or understood by application developers, and prone to errors. This also happens on Windows when static linking because we get no notification of threads exiting there.

nhorman replied:
agreed.  Given that this is an internal-only api we should be able to do so at our leisure in the future

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+     * NOTE: We use CRYPTO_THREAD_run_once here, rather than the
+     * RUN_ONCE macros.  We do this because this code is included both in
+     * libcrypto, and in fips.[dll|dylib|so].  FIPS attempts to avoid doing
+     * one time initalization of global data, and so supresses the definition
```
As an aside: The reason we had this restriction was because we had no easy way of cleaning up global data at the time (it can't be cleaned up on provider tear down, because the provider might be loaded into multiple libctxs). So we made it a rule that everything had to go into the libctx (no global data that needs to be cleaned up). However we actually do have a way of doing this now in the self_test `cleanup` function

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+        /*
+         * make sure to assign it to our master key thread-local storage
+         */
+        CRYPTO_THREAD_set_local(&master_key, mkey);
```
This call can fail. We should check it.

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+        /*
+         * Assign it to the appropriate entry in the master key sparse array
+         */
+        ossl_sa_MASTER_KEY_ENTRY_set(mkey, id, entry);
```
This can fail. We should check

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+     * that we've been provided.  Note we cast the pointer to a uintptr_t so
+     * as to use it as an index in the sparse array
+     */
+    ctxd = ossl_sa_CTX_TABLE_ENTRY_get(entry->ctx_table, (uintptr_t)ctx);
```
Should we get the concrete libctx here? Otherwise we risk confusion between `NULL` and the concrete default libctx.

paulidale commented:
> An alternative suggestion is to just create a completely lockless `OSSL_LIB_CTX`. All loading/unloading of providers must only be done before the libctx can be used by multiple threads. Probably we would also have to prefetch all the algorithms too. Once all the providers are loaded the libctx could be used like any other (including shared between threads), but no locks would be required.

IMO: this is the nice simple way forward...

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,451 @@
+        /*
+         * Assign to the entry in the table so that we can find it later
+         */
+        ossl_sa_CTX_TABLE_ENTRY_set(entry->ctx_table, (uintptr_t)ctx, ctxd);
```
Could fail

Sashan commented on crypto/initthread.c:
```diff
@@ -286,19 +307,15 @@ void *ossl_thread_event_ctx_new(OSSL_LIB_CTX *libctx)
 {
-    CRYPTO_THREAD_cleanup_local(tlocal);
-    OPENSSL_free(tlocal);
+    CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_TEVENT_KEY, ctx, NULL);
```
Then I still must be missing something here. At line 186 in `context_init()` we call `ossl_thread_event_ctx_new()` function which reads as follows:
```
289 int ossl_thread_event_ctx_new(OSSL_LIB_CTX *libctx)
290 {
291     THREAD_EVENT_HANDLER **hands = NULL;
292
293     hands = OPENSSL_zalloc(sizeof(*hands));
294     if (hands == NULL)
295         goto err;
296
297     if (!CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_TEVENT_KEY, libctx, hands))
298         goto err;
299
300     /*
301      * We should ideally call ossl_thread_register_fips() here. This function
302      * is called during the startup of the FIPS provider and we need to ensure
303      * that the main thread is registered to receive thread callbacks in order
304      * to free |hands| that we allocated above. However we are too early in
305      * the FIPS provider initialisation that FIPS_get_core_handle() doesn't work
306      * yet. So we defer this to the main provider OSSL_provider_init_int()
307      * function.
308      */
309
310     return 1;
311  err:
312     OPENSSL_free(hands);
313     return 0;
314 }
```
at line 293 we allocate `hands` which we then assign to thread local storage. I'm not able to figure out when/how the hands is getting freed. For example consider call to `ossl_fips_prov_ossl_ctx_new()` at line 189 in `context_init()` fails we then do `goto err` where we call `context_deinit_objs()` which in turn calls `ossl_thread_event_ctx_free()` which does just this:
```
316 void ossl_thread_event_ctx_free(OSSL_LIB_CTX *ctx)
317 {
318     CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_TEVENT_KEY, ctx, NULL);
319 }
```
And if I understand the code in `CRYPTO_THREAD_set_local_ex()` right setting the NULL will leak the `hands` we allocated during the _init phase. But I might be missing something.

nhorman replied:
This function (ossl_thread_event_ctx_new) is only called for the fips provider, when calling OSSL_LIB_CTX_new, and is freed when calling OPENSSL_thread_stop_ex on thread exit.  Note that ossl_ctx_thread_stop is defined differently for FIPS, where we call init_get_thread_local_ex, to get the hands function based on the calling thread and context pointer.  Thats where the hands array gets freed.

Sashan replied:
So in other words: we don't care about a leak in the error path, because error path in `context_init()` function does not call `ossl_ctx_thread_stop()`. I think I can live with it.

nhorman replied:
Hmm, yeah, thats a good point, we expect OPENSSL_thread_stop[_ex] to be called by the calling  thread prior to exiting, but if it cleans up the context in question prior to doing so, the thread stop handler won't find the pointer in the key lookup table to free, and we leak.

Thats no different than the prior method, in which we just use the OS based theread-local key, so I suppose its accurate to say we don't care about cleanup in the error path, but that is ugly.  

We could modify this such that ossl_thread_event_ctx_free just didn't exist (i.e. we leave the thread local pointer in place on cleanup), that would give OPENSSL_thread_stop_ex the opportunity to find the pointer and free it.  Thoughts?

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,408 @@
+ *
+ * NOTE: This differs from the CRYPTO_THREAD_[get|set]_local api set in that
+ * this api stores a single OS level thread-local key per-process, and manages
+ * subsequent keys using a series of sparse arrays stored against that aforementioned
```
arrays and sparse arrays

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,408 @@
+ *                                  per-<thread*ctx> data
+ *
+ * It uses sparse arrays to map:
+ *   - Thread-local key IDs to master key entries.
```
This is no longer a sparse array

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,408 @@
+    if (!CRYPTO_THREAD_run_once(&master_once, init_master_key))
+        return 0;
+
+    if (id >= CRYPTO_THREAD_LOCAL_KEY_MAX)
```
This perhaps should be an `ossl_assert`

mattcaswell commented on crypto/threads_common.c:
```diff
@@ -0,0 +1,408 @@
+     *
+     * Assign to the entry in the table so that we can find it later
+     */
+     ossl_sa_CTX_TABLE_ENTRY_set(mkey[id].ctx_table, (uintptr_t)ctx, data);
```
This can fail

mattcaswell reviewed (approved):
Looks good. One very picky comment below. But approved either way.

nhorman commented on crypto/threads_common.c:
```diff
@@ -56,9 +56,11 @@
- *   - Library context pointers to context-specific data.
+ * It uses the following lookup pattern:
+ *   1) A global os defined key to a per-thread fixed array
+ *   2) An os defined key id as an index to (1) to get a sparse array
```
yeah, that should say libcrypto defined key

Sashan commented on crypto/initthread.c:
```diff
@@ -83,64 +84,100 @@ static GLOBAL_TEVENT_REGISTER *get_global_tevent_register(void)
+
 static THREAD_EVENT_HANDLER **
-init_get_thread_local(CRYPTO_THREAD_LOCAL *local, int alloc, int keep)
+init_manage_thread_local(OSSL_LIB_CTX *ctx, int alloc, int keep)
```
this is a nit:
IMO using `manage_thread_local()` here would be a better option here. Single verb in function name leaves no doubts in what function is actually doing.

Sashan commented on crypto/initthread.c:
```diff
@@ -83,64 +84,100 @@ static GLOBAL_TEVENT_REGISTER *get_global_tevent_register(void)
-    long sane;
-    CRYPTO_THREAD_LOCAL value;
-} destructor_key = { -1 };
+static ossl_inline THREAD_EVENT_HANDLER **init_fetch_clear_thread_local(OSSL_LIB_CTX *ctx)
```
same here: the name `init_fetch_clear`_... it does not help me to understand what's going to happen unless I got and look under the hood.

Sashan reviewed (changes requested):
I would drop `init_` prefix from all function names. so `init_manage_thread_local()` becomes `manage_thread_local()`

init_fetch_clear_... should become clear_thread_local() etc, thank you.

openssl-machine commented:
This pull request is ready to merge

nhorman commented:
pushed to master, thank you!
